### PR TITLE
Remote file download progress

### LINKED ISSUE #2812: Feature request: Show remote_file progress
Hi,

Using `remote_file` on a large file results in a recipe that looks like it's stuck because nothing seems to be happening (I'm pulling down 546 Mb over http). It'd be very useful to show the progress as it downloads the file (assuming I haven't missed an option that makes this possible).

And thanks for making Chef!

### PR DESCRIPTION
**Pull request by thommay**
Based on @brandocorp's work in #3082 rebased, with some tweaks around configuration and tests.

cc @chef/client-core
Fixes #3082, Fixes #2812

### CODE REVIEW

File: lib/chef/http.rb [hunk 1]
```diff
@@ -74,6 +74,9 @@ def self.use(middleware_class)
     attr_reader :redirect_limit
 
     attr_reader :options
+    attr_reader :show_progress
```
Comment (coderanger) [inline lib/chef/http.rb]: Are these being used?

**Comment by thommay** [comment]
@coderanger nope, gone.

File: lib/chef/formatters/doc.rb [hunk 2]
```diff
@@ -258,6 +259,21 @@ def resource_action_start(resource, action, notification_type = nil, notifier =
...
+        @progress[resource] ||= 0
+
+        percent_complete = (current.to_f / total.to_f * 100).to_i
+
+        if percent_complete > @progress[resource]
+
+          @progress[resource] = percent_complete
+
+          if percent_complete % interval == 0
+            start_line " - Progress: #{percent_complete}%", :green
```
Comment (tyler-ball) [inline lib/chef/formatters/doc.rb]: Will this continually overwrite the same line, or will it output multiple lines for the `- Progress:`?

Comment (thommay) [inline lib/chef/formatters/doc.rb]: multiple lines; there's no real way to do the update in a cross platform way that I know of (it requires a lot of terminal ANSI codes and nasty to make it work)

**Comment by lamont-granquist** [comment]
@thommay https://github.com/jfelchner/ruby-progressbar ?

**Comment by thommay** [comment]
interesting, thanks. Not sure it's actually necessary though, either.

**Comment by lamont-granquist** [comment]
yeah, but it might be a nice way of leveraging curses stuff to make progressbars on the same line without having to mess with curses and ANSI codes yourself...

**Comment by stevendanna** [comment]
I don't think we need in-place updating right now, but in general if you are updating a single line and you are the only one writing to stdout, you don't need anything fancy. Write the line without a newline appending, write subsequent lines starting with a carriage return.

**Comment by danielsdeleo** [comment]
I'd really prefer not to have multiple lines, it makes the output too cluttered. If that's the only way we can do it then at least it should have some additional logic (like bucketing by 5% intervals or something) to reduce clutter.

File: lib/chef/formatters/doc.rb [hunk 3]
```diff
@@ -258,6 +259,21 @@ def resource_action_start(resource, action, notification_type = nil, notifier =
...
+      def resource_update_progress(resource, current, total, interval)
+        @progress[resource] ||= 0
+
+        percent_complete = (current.to_f / total.to_f * 100).to_i
+
+        if percent_complete > @progress[resource]
+
+          @progress[resource] = percent_complete
+
+          if percent_complete % interval == 0
```
Comment (coderanger) [inline lib/chef/formatters/doc.rb]: Doesn't this miss printing if it jumps from, say, 19% to 21%?

**Comment by coderanger** [comment]
:+1: for more bucketing, I would also do it by time, like only print an update every 5 seconds. That should cut down on useless log spam for small files or big files on internal gigabit.

File: lib/chef/formatters/doc.rb (hunk 2 above)
Comment (tyler-ball) [inline lib/chef/formatters/doc.rb]: :+1:

File: lib/chef/formatters/doc.rb (hunk 3 above)
Comment (brandocorp) [inline lib/chef/formatters/doc.rb]: Yes, I believe it would

**Comment by brandocorp** [comment]
@danielsdeleo, @coderanger 
The current behavior is to update at a user defined percentage-interval, with the default at 20%. This should limit the log spam from this to a maximum of 5 events per remote_file resource. Is this what you meant by 'bucketing'? 

Time based eventing could be done instead, but for very large files, or very slow transfers it could even more noisy.

@thommay 
I lost track of this one, but let me know if I can help!

**Comment by smurawski** [comment]
:+1:

**Comment by lamont-granquist** [comment]
:+1: good enough to ship as MVP, its off by default and we can make it prettier later...  :shipit:

File: lib/chef/formatters/doc.rb (hunk 3 above)
Comment (thommay) [inline lib/chef/formatters/doc.rb]: Yup, but I'm not sure it matters. I think this is fine as a best effort, just to give people some idea of what's happening for really big files.

**Comment by danielsdeleo** [comment]
Yeah, I think 10% is fine. And we can improve later, as well.

**Pull request merged**